service/src/test/java/uk/gov/hmcts/reform/fpl/service/ConfidentialDetailsServiceTest.java

```diff
@@ -1,130 +1,558 @@
+            List<Element<Respondent>> respondents = List.of(respondentWithConfidentialFields(ID, CONFIDENTIAL));
+
+            List<Element<Respondent>> confidentialRespondents = service.getConfidentialDetails(respondents);
+
+            assertThat(confidentialRespondents).containsOnly(respondentWithConfidentialFields(ID, NO_VALUE));
+        }
+
+        @Test
+        void shouldNotAddToListWhenRespondentsIsEmptyList() {
+            List<Element<Respondent>> confidentialRespondents = service.getConfidentialDetails(List.of());
+
+            assertThat(confidentialRespondents).isEmpty();
+        }
+
+        @Test
+        void shouldOnlyAddConfidentialDetailsToListOnlyWhenAnsweredYesForRespondent() {
+            List<Element<Respondent>> respondents = List.of(
+                respondentWithConfidentialFields(ID, CONFIDENTIAL),
+                respondentWithConfidentialFields(randomUUID(), NOT_CONFIDENTIAL));
+
+            List<Element<Respondent>> confidentialRespondents = service.getConfidentialDetails(respondents);
+
+            assertThat(confidentialRespondents).containsExactly(respondentWithConfidentialFields(ID, NO_VALUE));
+        }
+
+        @Test
+        void shouldRemoveConfidentialDetailsWhenMarkedAsConfidential() {
+            List<Element<Respondent>> respondent = List.of(respondentWithConfidentialFields(ID, CONFIDENTIAL));
+
+            assertThat(service.removeConfidentialDetails(respondent))
+                .containsOnly(respondentWithRemovedConfidentialFields(ID));
+        }
+
+        @Test
+        void shouldNotRemoveConfidentialDetailsWhenNotMarkedAsConfidential() {
+            List<Element<Respondent>> respondents = List.of(respondentWithConfidentialFields(ID, NOT_CONFIDENTIAL));
+
+            assertThat(service.removeConfidentialDetails(respondents)).isEqualTo(respondents);
+        }
+
+        @Test
+        void shouldAddConfidentialDetailsToCaseDetailsWhenClassExists() {
+            CaseDetails caseDetails = CaseDetails.builder().data(new HashMap<>()).build();
+            List<Element<Respondent>> respondents = List.of(respondentWithConfidentialFields(ID, CONFIDENTIAL));
+
+            service.addConfidentialDetailsToCase(caseDetails, respondents, RESPONDENT);
+
+            assertThat(caseDetails.getData()).containsKeys(RESPONDENT.getCaseDataKey());
+        }
+
+        @Test
+        void shouldRemoveConfidentialDetailsFromCaseDetailsWhenNoConfidentialDetails() {
+            Map<String, Object> data = new HashMap<>();
+            data.put(RESPONDENT.getCaseDataKey(), "");
+
+            CaseDetails caseDetails = CaseDetails.builder().data(data).build();
+
+            service.addConfidentialDetailsToCase(caseDetails, emptyList(), RESPONDENT);
+
+            assertThat(caseDetails.getData()).isEmpty();
+        }
+
+        @Test
+        void shouldAddEmptyElementWhenRespondentIsEmpty() {
+            List<Element<Respondent>> elements = service.combineRespondentDetails(
```

Comment:
we could use the unwrap elements method here if we wanted to remove some of the .get(0).getValue (not fussed if you want to leave though)

Reply:
yeah again, am not too fussed about test implementation here. It's like this in master. I just wanna focus on making the service generic in this PR

Reply:
happy with this :)